src/Model/Componentfactory/include/Componentfactory.hpp

```diff
@@ -329,8 +329,8 @@ namespace Model::Componentfactory {
         nlohmann::json &topology,
         std::vector<std::unique_ptr<Network::Node>> &nodes) const override {
       auto topology_schema = this->get_schema(true);
-      Aux::schema::validate_json(topology, topology_schema);
       Aux::schema::apply_defaults(topology, topology_schema);
+      Aux::schema::validate_json(topology, topology_schema);
```

Comment:
An alternative to the "required" removal of default attributes is validating after applying the defaults as done here. The advantage is that this might catch some errors, the disadvantage is that the json schema will have warnings about missing values which are fine since defaults exist.

Reply:
I think that not warning about the missing fields is better, but I'm not sure.

Reply:
Currently there is the "relax_schema" function 

```cpp
      auto component_schema = Aux::schema::relax_schema(
          component->get_schema(include_external),
          {"desired_delta_x", "number_of_stochastic_steps", "theta_P",
           "sigma_P", "theta_Q", "sigma_Q", "balancelaw", "scheme"});
      node_schemas[name] = Aux::schema::make_list_schema_of(component_schema);
```

which prevents errors. But that requires hardcoded attributes.

Reply:
How is it handled right now? I guess we either don't need this function anymore or we must provide a static method, that declares, which entries are defaultible and therefore not required.
Or, maybe easiest, we make all entries, except "id", "from" and "to" defaultible and therefore remove required from all of them.

Reply:
well at the time this is happening we already know the defaults so we could remove the requirements just from those

Reply:
af73156 solves this problem

Reply:
I don't like this solution for the following reason:
The defaults are a convenience feature and can be totally ignored or later changed.

I think it would be confusing, if I generate the schemas, later on decide that actually all my pipelines have the same `desired_delta_x`, put it into the default section but then the schema is invalidated in the `topology.json`, because I removed the individual `desired_delta_x` entries there.

Reply:
As soon as you regenerate the schemas that "missing" error goes away. And you can not avoid the schema regeneration if you want to have required values at all - I mean everything could have a default

Reply:
I mean the generated schema files themselves are also a convenience feature which can be totally ignored. And the actual schema validation is not impacted since the defaults will be applied beforehand

Reply:
Question is this: Would schemas be used, if you have to remember to update them, when you supply defaults?
I guess (without users its hard to know, hopefully this will change, once the relevant paper is published.) that people will generate the schemas before working on a problem and then fill the files.
Then they decide that defaults are a good idea and suddenly the schemas fail to validate.

I as a user would just not expect a schema to only be usable if some other json file contains a specific key, and that key may even change.
Therefore I think having everything except "id", "from","to" not-required is less surprising.
Alternatively have everything required.
In that case the topology.json is invalid whenever defaults are used but at least the schema doesn't depend on entries in the jsons themselves.

Reply:
A possibly involved solution would be to reference the problem_data.json with `$ref` and then apply a conditional schema:
https://opis.io/json-schema/2.x/conditional-subschemas.html#:~:text=if-then-else,-This%20is%20a&text=Every%20keyword%20value%20must%20be,be%20at%20the%20same%20time).

In that way we would force once more the file structure we have but I'm ok with that.

Reply:
I honestly don't see how someone would fail to work with this - every way one could think of works:
- turning it off and on again works (deleting subschemas and regenerating them)
- reading the docs on defaults works (they would tell you to recreate the schemas)
- deleting the key from the required list in the schema manually works
- googling this problem/asking how to deal with it would work
- simply ignoring the warning works (this is basically the same as if we kept everything as "required" ("Alternatively have everything required.")

I really don't see how this is worse than having everything be required and getting tons of warnings, or getting no schema help at all and failing at the "run" step. Plus it is extremely discoverable. If you try to figure out defaults for the first time you are going to look for that in the documentation and the fact that schemas will break will be right there basically impossible to miss. And if you actually miss it then it is immediately obvious what action caused the warning and you are going to read the documentation again a bit more carefully and immediately spot the issue.

That being said, trying to make this work all the time with `$ref` scratches my perfectionist itch even though I think that this probably introduces unnecessary complexity making the schemas even less readable than they are already and causing errors to be less readable too (making the unit regex error messages somewhat readable was also a challenge). And I am also not quite sure how this `$ref` thing would work, since it is usually meant to include schema snippets in some place and not actual json. Using `$ref` inside the data jsons would work, but that is user generated so not what we want. And the data document states what schema applies to it, I am not sure the schema has a say as to what it is applied to (i.e. not only the data json but also the defaults). But this might somehow be possible (although I doubt it would be very readable/maintainable) 🤷

Reply:
Well then, we'll choose your way.
Then please write a section on schema validation into the userguide.tex.
You can find some inspiration (though not much) in
https://arxiv.org/abs/2108.00687

That's the paper that hopefully gets published to draw attention to grazer.

Reply:
But don't do that on this branch.
I think, I'll merge this as it is and you can write the docs on a second PR. I don't expect merge conflicts on the userguide.tex and that way I can get on with the other PRs.